grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcHealthCheckService.java

```diff
@@ -0,0 +1,293 @@
+        String service = request.getService();
+        final Builder builder = HealthCheckResponse.newBuilder();
+        try {
+            builder.setStatus(checkServingStatus(service));
+        } catch (StatusException ex) {
+            builder.setStatus(ServingStatus.SERVICE_UNKNOWN);
+        }
+        HealthCheckResponse response = builder.build();
+        responseObserver.onNext(response);
+        if (watchers.containsKey(service)) {
+            watchers.get(service).put(responseObserver, response.getStatus());
+        } else {
+            final IdentityHashMap<StreamObserver<HealthCheckResponse>, ServingStatus> mp =
+                    new IdentityHashMap<>();
+            mp.put(responseObserver, response.getStatus());
+            watchers.put(service, mp);
+        }
+        ((ServerCallStreamObserver<HealthCheckResponse>) responseObserver).setOnCancelHandler(() -> {
+            if (watchers.containsKey(request.getService())) {
+                watchers.get(service).remove(responseObserver);
+                if (watchers.get(service).isEmpty()) {
+                    watchers.remove(service);
+                }
+            }
+        });
+    }
+
+    void serviceAdded(ServiceConfig cfg) {
+        if (server != null) {
+            if (server != cfg.server()) {
+                throw new IllegalStateException("Cannot be added to more than one server");
+            } else {
+                return;
+            }
+        }
+
+        server = cfg.server();
+        server.addListener(new ServerListenerAdapter() {
+            @Override
+            public void serverStarted(Server server) {
+                serverHealth.setHealthy(true);
+            }
+
+            @Override
+            public void serverStopping(Server server) {
+                serverHealth.setHealthy(false);
+            }
+        });
+    }
+
+    @VisibleForTesting
+    ServingStatus checkServingStatus(String serverName) throws StatusException {
+        if (serverName.isEmpty()) {
+            if (isHealthy()) {
+                boolean isHealthyForAllGrpcService = true;
+                for (HealthChecker healthChecker : grpcServiceHealthCheckers.values()) {
+                    if (!healthChecker.isHealthy()) {
+                        isHealthyForAllGrpcService = false;
+                    }
+                }
+                if (isHealthyForAllGrpcService) {
+                    return ServingStatus.SERVING;
+                } else {
+                    return ServingStatus.NOT_SERVING;
+                }
+            } else {
+                return ServingStatus.NOT_SERVING;
+            }
+        } else if (grpcServiceHealthCheckers.containsKey(serverName)) {
+            if (grpcServiceHealthCheckers.get(serverName).isHealthy()) {
+                return ServingStatus.SERVING;
+            } else {
+                return ServingStatus.NOT_SERVING;
+            }
```

Comment:
> It would be better to use AND condition to see check the status of a service.

It seems to make sense.

> How about returning a ServingStatus and throwing an exception later in the caller site?

It seems good.